File: DNA.VisualBasic/DNA.VisualBasic/OData/ODataEnableQueryCodeFix.vb
Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.CodeActions
Imports Microsoft.CodeAnalysis.CodeFixes
Imports Microsoft.CodeAnalysis.CodeRefactorings
Imports Microsoft.CodeAnalysis.Editing
Imports Microsoft.CodeAnalysis.Rename
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

<ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=NameOf(ODataEnableQueryCodeFix))>
Public Class ODataEnableQueryCodeFix
    Inherits CodeFixProvider

    ' TODO: Replace with actual diagnostic id that should trigger this fix.
    Public Const DiagnosticId As String = "DNA101"

    Public NotOverridable Overrides ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String)
        Get
            Return ImmutableArray.Create(DiagnosticId)
        End Get
    End Property

    Public NotOverridable Overrides Function GetFixAllProvider() As FixAllProvider
        Return WellKnownFixAllProviders.BatchFixer
    End Function

    Public NotOverridable Overrides Async Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
        Dim root = Await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(False)

        ' TODO: Replace the following code with your own analysis, generating a CodeAction for each fix to suggest

        Dim diagnostic = context.Diagnostics.First()
        Dim diagnosticSpan = diagnostic.Location.SourceSpan

        Dim node = TryCast(root.FindNode(context.Span), MethodStatementSyntax)

        context.RegisterCodeFix(
        CodeAction.Create("Add <EnableQuery>",
                          Function(c) EnableQueryAsync(context.Document, node, c)), diagnostic)
    End Function

    Private Async Function EnableQueryAsync(document As Document, node As MethodStatementSyntax,
                                 cancellationToken As CancellationToken) As Task(Of Document)

        Dim root = Await document.GetSyntaxRootAsync

        Dim newMethod = node.WithAttributeLists(node.AttributeLists.Add(SyntaxFactory.AttributeList(
                    SyntaxFactory.SingletonSeparatedList(Of AttributeSyntax)(
                        SyntaxFactory.Attribute(
                            SyntaxFactory.IdentifierName("EnableQuery"))))))

        Dim newRoot = root.ReplaceNode(node, newMethod)
        Dim newDocument = document.WithSyntaxRoot(newRoot)
        Return newDocument
    End Function
End Class

File: DNA.CSharp/DNA.CSharp/OData/ODataEnableQueryCodeFix.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Rename;
using Microsoft.CodeAnalysis.Text;

namespace DNA.CSharp.OData
{
    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(ODataEnableQueryCodeFix))]
    public class ODataEnableQueryCodeFix : CodeFixProvider
    {
        // TODO: Replace with actual diagnostic id that should trigger this fix.
        public const string DiagnosticId = "DNA101";

        public sealed override ImmutableArray<string> FixableDiagnosticIds
        {
            get { return ImmutableArray.Create(DiagnosticId); }
        }

        public sealed override FixAllProvider GetFixAllProvider()
        {
            return WellKnownFixAllProviders.BatchFixer;
        }

        public sealed async override Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);

            var diagnostic = context.Diagnostics.First();
            var diagnosticSpan = diagnostic.Location.SourceSpan;

            var node = root.FindNode(context.Span) as MethodDeclarationSyntax;

            context.RegisterCodeFix(
            CodeAction.Create("Add [EnableQuery]",
                              c => EnableQueryAsync(context.Document, node, c)), diagnostic);
        }

        private async Task<Document> EnableQueryAsync(Document document, MethodDeclarationSyntax node,
                             CancellationToken cancellationToken)
        {
            var root = await document.GetSyntaxRootAsync();
            var newMethod = node.WithoutLeadingTrivia().WithAttributeLists(node.AttributeLists.Add(SyntaxFactory.AttributeList(
            SyntaxFactory.SingletonSeparatedList<AttributeSyntax>(
                SyntaxFactory.Attribute(
                    SyntaxFactory.IdentifierName("EnableQuery")))))).WithLeadingTrivia(node.GetLeadingTrivia());
            var newRoot = root.ReplaceNode(node, newMethod);
            var newDocument = document.WithSyntaxRoot(newRoot);
            return newDocument;
        }

    }
}